Add nesting-roast-half-aggregation-adaptor-signatures review to chaincode-podcast (#650)
bitcointranscripts/bitcointranscripts: chaincode-podcast/nesting-roast-half-aggregation-adaptor-signatures.md

Reviewer: `Speaker 0` _**might**_ be Pieter. Please identify the speakers.

Pieter: 00:19:11

You can download the audio embedded in [spotify webpage](https://creators.spotify.com/pod/profile/chaincode/episodes/Pieter-Wuille-and-Tim-Ruffing---Nesting--ROAST--Half-Aggregation--Adaptor-Signatures-part-2-e1sdgjf). 
https://d3ctxlq1ktw2nl.cloudfront.net/staging/2022-11-18/1e2ca48f-a779-0eda-2bbd-357c9a24e294.mp3

Reference part 1:
https://btctranscripts.com/chaincode-podcast/schnorr-musig-frost-and-more
https://d3ctxlq1ktw2nl.cloudfront.net/staging/2022-11-16/21bed2ba-2527-8c4f-b406-cdb955f43061.mp3
Mark Erhart(Host)
Adam Jonas(Host)

---
title: 'Nesting, ROAST, Half-Aggregation, Adaptor Signatures (Part 2)'
transcript_by: '0xmilarepa via review.btctranscripts.com'
media: 'https://podcasters.spotify.com/pod/show/chaincode/episodes/Pieter-Wuille-and-Tim-Ruffing---Nesting--ROAST--Half-Aggregation--Adaptor-Signatures-part-2-e1sdgjf'
date: '2022-12-27'
tags:
  - 'adaptor-signatures'
  - 'bip32'
  - 'cisa'
  - 'coinjoin'
  - 'multisignature'
  - 'musig'
  - 'schnorr-signatures'
  - 'segwit'
  - 'signature-aggregation'
  - 'taproot'
speakers:
  - 'Pieter Wuille'
  - 'Tim Ruffing'
  - 'Adam Jonas'
  - 'Mark Erhart'
categories:
  - 'podcast'
aliases:
  - '/chaincode-labs/chaincode-podcast/nesting-roast-half-aggregation-adaptor-signatures/'
---
Tim Ruffing: 00:00:00

> Just as a warning: don't do this at home.

Reviewer: Please revert. Also, wrong grammar.  A **colon** should only be used after a complete, independent clause.

  - 'podcast'
aliases:
  - /chaincode-labs/chaincode-podcast/nesting-roast-half-aggregation-adaptor-signatures/
  - '/chaincode-labs/chaincode-podcast/nesting-roast-half-aggregation-adaptor-signatures/'
---
Speaker 1: 00:00:00
Tim Ruffing: 00:00:00

Just as a warning, so don't do this at home.
Just as a warning: don't do this at home.

Speaker 0: 00:00:01
Pieter Wuille: 00:00:01

Yes, of course.
Still have your friendly neighborhood cryptographer have a look at it.

Speaker 2: 00:00:16
Adam Jonas: 00:00:16

This is the second half of the conversation with Tim and Peter.

> If you have not listened to the first half, I'd suggest going back and listening to that episode.

Reviewer: Revert the comma, please

> If you have not listened to the first half I'd suggest going back and listening to that episode.

Reviewer: The original transcript actually included the **comma** correctly. Why was it removed?

Let's focus on that first.

Tim Ruffing: 00:08:25

MuSig2 was built with this nesting idea in mind in the sense that, but we didn't include it in the paper, but we think we have a way to do it or we thought we have a way to do it now, like a month ago, we discovered it doesn't really work.
So it's still an open research question.
It's not like, if you look at the scheme, you could think you could do it today, but please, please don't do it because it's still an open problem.
We have to think about it carefully, like maybe add some restrictions and then write a proper security proof for it.

Pieter Wuille: 00:08:55

When you say a way to do it that both encompasses what the actual algorithm is but also how do we go about proving it?
Because today...

Tim Ruffing: 00:09:04
Security and functionality, yeah.

Pieter Wuille: 00:09:06

Right, because you... MuSig2 is specifically designed to support nesting, originally designed to support nesting.

> So it, If you just look at the spec and do the naive thing...

Reviewer: ```suggestion
So if you just look at the spec and do the naive thing...
```

Tim Ruffing: 00:09:20

It is a natural thing you could do to make nesting work in a sense that you can create a signature and the signature would verify, but it doesn't mean that this specific way is secure, and this is the problem we still need to solve.

Pieter Wuille: 00:09:35

While at the same time if you try to do the same thing with the three-round MuSig1, either you'll conclude that you just can't make it work or you're gonna change the scheme in a way that will obviously and completely break it.

Tim Ruffing: 00:09:49

Yeah, yeah.
And this is what I was mentioning, this is really just only about the MuSig inside MuSig case, so not even talking about FROST inside MuSig, which will probably...

Pieter Wuille: 00:10:01

> Or with BIP32 on top, or with Taproot on top, or cross-input second iteration on top.

Reviewer: ```suggestion
Or with BIP32 on top, or with Taproot on top, or cross-input signature aggregation on top.
```

And I think this is where it's going to be used.
It's probably ROAST is nothing that you would use in your 2-of-3 at home, or even like a 3-of-5 at home or something like that.

Pieter Wuille: 00:17:56

But maybe because for your 2-of-3, you really just don't want to use a threshold signature scheme in the first place.

Tim Ruffing: 00:18:02

That's why I mentioned also 3-of-5.
Like even, let's say you have a 3-of-5 at home, I guess you wouldn't need ROAST because it's like, okay, I picked those three hardware wallets to make a signature or some of it failed, replace the one that failed with another one and try again.
You could do this restarting of FROST sessions basically manually.
So ROAST really, I think, is helpful when you have some federation of nodes around the world that are running automatically.
They're supposed to produce a signature every one minute or every ten minutes, and then you need the automatic way of making sure that they really can produce a signature.
I think this is where ROAST is strong.
But it's good for these federation use cases, but I think this is really like a niche case in a sense.

## Cross-input Signature Aggregation

Adam Jonas: 00:18:49

> So Peter, you have mentioned cross-input aggregation as the inspiration for a lot of the work that actually has been done to date.

Reviewer: ```suggestion
So Pieter, you have mentioned cross-input aggregation as the inspiration for a lot of the work that actually has been done to date.
```

Pieter Wuille: 00:20:32

Or you say the message is: key one signs message this, key two signs message that, key three signs message that, and everybody signs that.
That makes an interactive aggregate signature scheme, like just take a multi-signature scheme and turn into that and you're there.

Tim Ruffing: 00:20:49

Modulo some subtle details.
But, yeah, let's not go into that.
But just as a warning, don't do this at home.

Pieter Wuille: 00:20:56

Yes, of course.
Still have your friendly neighborhood cryptographer have a look at it.

Adam Jonas: 00:21:04

Before you move on, that's just because of being able to actually tease out private keys based on reused nonces or what's the actual issue that you'd be most concerned.
If I did that, what would be the first mistake I would make?

Pieter Wuille: 00:21:21

In the MuSig paper, I remember Russell O'Connor came up with this fairly far-fetched attack of, if you just do that, what I just said, and try to use that specifically in the context of Bitcoin transactions with one signature... there was a problem but I don't remember.

> Tim Ruffing: 00: 21:41

Reviewer: ```suggestion
Tim Ruffing: 00:21:41
```

There could be a problem if you're a single party and you have multiple messages that you want to sign.

> You can be tricked into, and you participate multiple times, you can be tricked into signing the same one instead of different ones.

Reviewer: ```suggestion

Pieter Wuille: 00:21:50

You can be tricked into, and you participate multiple times, you can be tricked into signing the same one instead of different ones.

```

And really the primitive we are talking about here, the verification side of it would take multiple public keys, multiple messages, where they all could be the same transaction but a little bit different, okay?
But only one signature.
So really this is a different interface already.
You couldn't just do this with the current Schnorr verification algorithm that we have in consensus code right now.

Pieter Wuille: 00:27:07

And I think a more fundamental reason, like today you cannot spend any input without a signature, assume it has a public key.
That would be a problem if you could spend an input without a signature.
Yet, if we go to a cross input signature aggregation world, a transaction with two inputs, well, there's only going to be one signature.
That means there's at least one input without a signature.
Of course, the idea is that signature will cover all of them simultaneously, but the rules cannot think about that.
Any kind of cross-input aggregation scheme is going to require an additional consensus soft fork rule.

Adam Jonas: 00:27:49

The community didn't feel like it was worth waiting to more fully baked cross-input?

Pieter Wuille: 00:27:54

> Yeah, I think, so when the discussions around Taproot, which Schnorr signatures started, there were a whole lot of ideas and many of them were actively being discussed because there were improvements to Taproot: there was Graftroot, there was Groot, and so forth.

Reviewer: ```suggestion
Yeah, I think, so when the discussions around Taproot, which Schnorr signatures started, there were a whole lot of ideas and many of them were actively being discussed because there were improvements to Taproot: there was Graftroot, there was g'root, and so forth.
```

Reviewer: https://bitcoinops.org/en/newsletters/2021/10/27/
https://bips.dev/341/

And how those would interact with signature aggregation is kind of unclear because there is a complication here, which is soft fork compatibility.
So we want the property obviously that a change that introduces cross-input aggregation is a soft fork, and also that things that could be built on top, extensions to the script language later, are a soft fork with respect to signature aggregation already existing.
And so this isn't a fundamental problem, but it's kind of annoying.

> Say, imagine there's an opcode change that introduces something like an opif, just something that changes the execution path through a single script.

Reviewer: ```suggestion
Say, imagine there's an opcode change that introduces something like an OP_IF, just something that changes the execution path through a single script.
```

You need to make sure that old nodes and new nodes agree on what signatures are being aggregated, even though some of them may execute the checksig that's being skipped and others that don't.
So if you think about it, it's non-trivial.

> One possibility is doing, well, whenever a soft fork is introduced that changes which codes are executed, you create a new separate batch for aggregated signatures, you just do the aggregated signatures that are according to the pre-soft fork nodes and then everything that's added or changed with respect to that, they go into a new bucket and you end up with two signatures: one for all the signatures, or one aggregated signature for all the things visible to old nodes, and then another one for the additional ones that are visible to new nodes.

Reviewer: ```suggestion
One possibility is doing, well, whenever a soft fork is introduced that changes which opcodes are executed, you create a new separate batch for aggregated signatures, you just do the aggregated signatures that are according to the pre-soft fork nodes and then everything that's added or changed with respect to that, they go into a new bucket and you end up with two signatures: one for all the signatures, or one aggregated signature for all the things visible to old nodes, and then another one for the additional ones that are visible to new nodes.
```

And now we have multi-signature schemes, and you said like the math will be pretty similar but it's actually, if you look at the specifics of this, you probably wouldn't want to use something based on, let's say, MuSig, or MuSig1, or MuSig2, and any of these.

Pieter Wuille: 00:31:45

Because they're kind of overkill for...

Tim Ruffing: 00:31:49

Kind of overkill, yeah.
What you actually need is a little bit weaker.
And also there are other issues when it comes to compatibility with batch verification or with batch validation.
One way to think about this is, in a signature aggregation scheme, it's a little bit like a multi-signature scheme, but as you said, the verifier will do the key aggregation because the verifier gets all the individual public keys and somehow would need to aggregate those keys.
And if we do this in a MuSig style, this key aggregation, then this key aggregation again is an elliptic curve computation, but we couldn't add this to the batch for doing batch validation.
So we couldn't batch it together with a signature validation or with a Taproot commitment openings.

> So probably there are other signature aggregation schemes more different from MuSig that would actually allow for fetch validation.

Reviewer: ```suggestion
So probably there are other signature aggregation schemes more different from MuSig that would actually allow for batch validation.
```

Reviewer: I couldn't find any reference to `fetch validation` as a technical term in this context.  Listen at 32:50

Today, but we should at least keep thinking about this.
This is a question that I often get.

Pieter Wuille: 00:50:45

It's a good one.

Adam Jonas: 00:50:47

Well, thank you both for giving us almost two hours and I will say it's been a pleasure to watch you both escape to different rooms and draw indecipherable things on whiteboards.
So thanks for coming in Tim, spending the week with us.

Mark Erhart: 00:51:05

All right, what did you think about that conversation?

Adam Jonas: 00:51:08

I thought it was great.
Not to toot my own horn, but I thought it was a great chance to pull a lot of info out of those two.
And I think the conversation about diving into the advanced signing protocols and multi-sigs versus threshold signatures and state of FROST and ROAST and MuSig2, I thought that was all really fun.
But thinking about taking advantage of batch verification and signature aggregation and its future, I thought was really quite great.
We talked about signature aggregation and the interest in that with Josie a couple episodes ago, but it gave me a chance to actually better understand it.

Mark Erhart: 00:51:43

Right.

> What you mentioned now was that you also thought it was interesting to see how the sausages made it a little more.

Reviewer: ```suggestion
What you mentioned now was that you also thought it was interesting to see how the sausage is made a little more.
```

About when to roll crypto.

Adam Jonas: 00:51:53

Well, I just like the conversation about them referring to MuSig1 as broken and just like being very upfront and public about the fact that it's broken and the upgrades that they made and sort of like the shower thoughts that Tim had to make MuSig2 work and all those things.
It's, I think it's fun.
So yeah, I hope you enjoyed listening to the episode as much as I enjoyed it.

Mark Erhart: 00:52:15

Yeah, well, we hope to have one soon for you again.

Adam Jonas: 00:52:18

I don't know, I don't know if we're gonna make it before our New Year's, but if not, hope you enjoy the holidays and Happy New Year's, and if so, we'll talk to you before then.

> Mark Erhart: 00:52:27

Reviewer: @0tuedon I suspect the `Speaker` in the final section isn't **Mark 'Murch' Erhardt** since Murch seems to have a bit of a stronger accent in **Part 1**.  
So, **Part 2** from Line 806 to 838 might be **another** host, or perhaps **Pieter Wuille** (works at Chaincode ?).  I’m unfamiliar with their voices to be certain. Jonas might be able to confirm. I’ll leave it to you to decide.

Reference **part 1**:
https://btctranscripts.com/chaincode-podcast/schnorr-musig-frost-and-more
https://d3ctxlq1ktw2nl.cloudfront.net/staging/2022-11-16/21bed2ba-2527-8c4f-b406-cdb955f43061.mp3
Mark Erhart(Host): 00:00:00
Adam Jonas(Host)

Reviewer 2: Is it for the bye? both said bye both Mark and Jonas

Reviewer 2: Oh i see I will check who that speaker is, @satsie

Reviewer: All the **Mark Erhart** in the last section from Line 806 to 838.

Author: all comments solved, i'll just wait for confirmation regarding the speaker of the last section.
for what it's worth, after listening to part 1, it seems to me that the voice of the second host who's speaking to Adam Jonas at the end matches that of Mark Erhardt's.
Also, i noticed part 1 spells his name as "Erhart", although, after double checking, i found his name is spelled "Erhardt". https://bitcoinops.org/en/about/

Reviewer 3: I think Murch (Mark) is the one speaking in this wrap up section at the end with Jonas. I agree that some of the lines don't sound like him but the comment about how the sausage is made is undeniably him. It also makes sense to me that he would be doing this conclusion portion with Jonas.

Reviewer: Just came across this. You guys are correct.
https://podcast.chaincode.com/about.html

Mark “Murch” **Erhardt** cohosts the podcast ....

date: '2022-12-27'
tags:
  - 'adaptor-signatures'
  - 'bip32'
  - 'cisa'
  - 'coinjoin'
  - 'multisignature'
  - 'musig'
  - 'schnorr-signatures'
  - 'segwit'
  - 'signature-aggregation'
  - 'taproot'
speakers:
  - 'Pieter Wuille'
  - 'Tim Ruffing'
  - 'Adam Jonas'
  - 'Mark Erhardt'
categories:
  - 'podcast'
aliases:
  - '/chaincode-labs/chaincode-podcast/nesting-roast-half-aggregation-adaptor-signatures/'
---
Tim Ruffing: 00:00:00

Just as a warning, don't do this at home.

Pieter Wuille: 00:00:01

Yes, of course.
Still have your friendly neighborhood cryptographer have a look at it.

Adam Jonas: 00:00:16

This is the second half of the conversation with Tim and Pieter.
If you have not listened to the first half, I'd suggest going back and listening to that episode.
We cover all sorts of fun things, including when to roll your own cryptography, why we prefer Schnorr signatures over ECDSA, Schnorr efficiency improvements, Multi-Sig, MuSig, interactive versus non-interactive signing, FROST, and more.
So, we're going to pick things back up talking about nesting.

> We'll cover ROAST, block-wide aggregation, adaptive signatures, atomic swaps.

Reviewer: ```suggestion
We'll cover ROAST, block-wide aggregation, adaptor signatures, atomic swaps.
```

And this is a way where you can, in some sense, get the best of both worlds, in the sense that if everyone is online and willing to sign, then the first first session that you would start would complete, and you get this nice two-round property.
But if you are under attack or maybe some of the signers is offline or some of the signers are offline, then you spend a few more rounds, but you can still run the... like get the signature within a reasonable amount of time.
This is the main idea of ROAST, or the main thing it achieves.

Adam Jonas: 00:16:38

And so, in terms of applicability of both FROST and ROAST, where are these things being deployed?
Where are they being used?
What are the applications you imagine?

Tim Ruffing: 00:16:48

Pieter has already mentioned that FROST already is maybe more a niche thing than MuSig because it's really only helpful when you really specifically need the threshold property.
And then if you add ROAST on top of it, I think ROAST is really helpful in settings where you not only need specifically the threshold property, but where you also need a large threshold setup.

> At Blockstream we have this liquid sidechain which is run by a federation and currently has an 11-of-15 setup.

Reviewer: ```suggestion
At Blockstream we have this Liquid sidechain which is run by a federation and currently has an 11-of-15 setup.
```

Pieter Wuille: 00:17:56

But maybe because for your 2-of-3, you really just don't want to use a threshold signature scheme in the first place.

Tim Ruffing: 00:18:02

That's why I mentioned also 3-of-5.
Like even, let's say you have a 3-of-5 at home, I guess you wouldn't need ROAST because it's like, okay, I picked those three hardware wallets to make a signature or some of it failed, replace the one that failed with another one and try again.
You could do this restarting of FROST sessions basically manually.
So ROAST really, I think, is helpful when you have some federation of nodes around the world that are running automatically.
They're supposed to produce a signature every one minute or every ten minutes, and then you need the automatic way of making sure that they really can produce a signature.
I think this is where ROAST is strong.
But it's good for these federation use cases, but I think this is really like a niche case in a sense.

## Cross-input Signature Aggregation

Adam Jonas: 00:18:49

So Pieter, you have mentioned cross-input aggregation as the inspiration for a lot of the work that actually has been done to date.

> Let's talk about signature aggregation, maybe start with interactive floor aggregation and sort of the cross-input aggregation history, and then we'll explore some other ways.

Reviewer: ```suggestion
Let's talk about signature aggregation, maybe start with interactive full aggregation and sort of the cross-input aggregation history, and then we'll explore some other ways.
```

Speaker 2: 00:00:16

This is the second half of the conversation with Tim and Peter.
If you have not listened to the first half I'd suggest going back and listening to that episode.
We cover all sorts of fun things, including when to roll your own cryptography, why we prefer Schnorr signatures over ECDSA, Schnorr efficiency improvements, Multi-Sig, MuSig, interactive versus non-interactive signing, FROST, and more.
So, we're going to pick things back up talking about nesting.
We'll cover ROAST, block-wide aggregation, adaptive signatures, atomic swaps.
This is a great conversation and I hope you enjoy it as much as I enjoyed recording it.

## Nesting

Speaker 2: 00:00:50

So FROST sounds exciting, especially when used in combination with MuSig, there's this sort of idea for the future called nesting.
How could these ideas be combined?

Speaker 1: 00:01:11

> Yeah, so again, MuSig is an event setup and FROST is a t-event setup, where you just require some subset, but now you can think of combining or nesting those in a three-style fashion.

Reviewer: ```suggestion
Yeah, so again, MuSig is a n-of-n setup and FROST is a t-of-n setup, where you just require some subset, but now you can think of combining or nesting those in a three-style fashion.
```

So it's basically a 2-of-2 MuSig at the top and on my side it's a 2-of-3 on that side of the tree.

Speaker 0: 00:02:20

Yeah, it's like a 2-of-2 of a 2-of-3 and a single key.

Speaker 1: 00:02:25

Yeah, and you're on your side and now you can go a step further.
You shouldn't be even aware of the fact that I use a 2-of-3 under the hood.
First, because of privacy, maybe I just don't want to reveal it, but also just for simplicity, because the Lightning protocol spec shouldn't be concerned with what I do locally with my keys.
So I don't want to reveal that I use a 2-of-3, not only for privacy reasons, but also to keep the specification simple because then whenever, let's say in the future, I want to do something else with my keys for every new use case in a sense, or a key setup case, you would need to change the Lightning specification.
That's not what we want to do.
So this is the rough application idea of nesting.
In this example, it would be a FROST inside the MuSig, but you could also think of arbitrary combinations, like a MuSig inside the MuSig, MuSig inside the FROST, and arbitrary trees you can build.

Speaker 0: 00:03:22

> And in addition, we have other things too, like we have BIP32 derivation, like can you do a FROST setup between participants and then get the next pub out that you can derive multiple keys from that all of them now may be signable.

Reviewer: ```suggestion
And in addition, we have other things too, like we have BIP32 derivation, like can you do a FROST setup between participants and then get the xpub out that you can derive multiple keys from that all of them now may be signable.
```

https://trezor.io/learn/supported-assets/bitcoin/what-is-a-public-key-xpub?srsltid=AfmBOopRnFZVlTP4J5riGb8j080_NZ125aPdbESUeY81lfVoJ7uEH3JB

Speaker 0: 00:08:55

When you say a way to do it that both encompasses what the actual algorithm is but also how do we go about proving it?
Because today...

Speaker 1: 00:09:04
Security and functionality, yeah.

Speaker 0: 00:09:06

Right, because you... MuSig2 is specifically designed to support nesting, originally designed to support nesting.
So it, If you just look at the spec and do the naive thing...

Speaker 1: 00:09:20

It is a natural thing you could do to make nesting work in a sense that you can create a signature and the signature would verify, but it doesn't mean that this specific way is secure, and this is the problem we still need to solve.

Speaker 0: 00:09:35

While at the same time if you try to do the same thing with the three-round MuSig1, either you'll conclude that you just can't make it work or you're gonna change the scheme in a way that will obviously and completely break it.

Speaker 1: 00:09:49
Yeah, yeah.
And this is what I was mentioning, this is really just only about the MuSig inside MuSig case, so not even talking about FROST inside MuSig, which will probably...

Speaker 0: 00:10:01

> Or with BEP32 on top, or with Taproot on top, or cross-input second iteration on top.

Reviewer: ```suggestion
Or with BIP32 on top, or with Taproot on top, or cross-input signature aggregation on top.
```

And really the primitive we are talking about here, the verification side of it would take multiple public keys, multiple messages, where they all could be the same transaction but a little bit different, okay?
But only one signature.
So really this is a different interface already.
You couldn't just do this with the current Schnorr verification algorithm that we have in consensus code right now.

Speaker 0: 00:27:07

And I think a more fundamental reason, like today you cannot spend any input without a signature, assume it has a public key.
That would be a problem if you could spend an input without a signature.
Yet, if we go to a cross input signature aggregation world, a transaction with two inputs, well, there's only going to be one signature.
That means there's at least one input without a signature.
Of course, the idea is that signature will cover all of them simultaneously, but the rules cannot think about that.
Any kind of cross-input aggregation scheme is going to require an additional consensus soft fork rule.

Speaker 2: 00:27:49

The community didn't feel like it was worth waiting to more fully baked cross-input?

Speaker 0: 00:27:54

> Yeah, I think, so when the discussions around Taproot, which North Signature started, there were a whole lot of ideas and many of them were actively being discussed because there were improvements to Taproot: there was Graftroot, there was Groot, and so forth.

Reviewer: Check?

Author: I corrected "North Signature" for "Schnorr signatures", but not so sure about Groot. I listened several times and Pieter seems to be actually saying "Groot", although i've never hear of such a name for a BIP.

And half aggregation, if now this signature that's published on the blockchain would be half aggregated, then it's not the full signature that would be published there.
We would exactly remove that part that you need to look at to make the second transaction happen.
So this is how this would interfere.
And now, as you say, it could actually be done with ECDSA.
So one very simple, of course not very satisfactory solution, would be, okay, if you want to run an adaptor signature protocol, then resort to ECDSA signatures.
So one, maybe a little bit more clever approach would be, okay, you could do it with Schnorr signatures but maybe have a marker in your transaction that says okay this could be aggregated or this could not be aggregated.
Of course this...

Speaker 0: 00:42:39

That's unfortunate because now you're revealing to the chain that this is really a data carrying signature and...

Speaker 1: 00:42:47

This is not good for privacy, obviously.

Speaker 0: 00:42:49

> And of course adaptor signatures are an alternative to say, HTLCs, which are used for exactly the same purpose today, except they reveal hash per image, rather than using the signature itself as the data channel.

Reviewer: Check `hash per image` ?

Speaker 1: 00:49:07

I agree, yeah, and we haven't really touched upon the status of the specifications.

> I think the MuSig2 pip, as you're saying, is in a good shape, I think.

Reviewer: ```suggestion
I think the MuSig2 BIP, as you're saying, is in a good shape, I think.
```
BIP (Bitcoin Improvement Proposal) ?
